hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3AccessGrantConfiguration.java

```diff
@@ -0,0 +1,108 @@
+ */
+
+package org.apache.hadoop.fs.s3a;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.test.AbstractHadoopTestBase;
+import org.junit.Assert;
+import org.junit.Test;
+
+import software.amazon.awssdk.awscore.AwsClient;
+import software.amazon.awssdk.s3accessgrants.plugin.S3AccessGrantsIdentityProvider;
+import software.amazon.awssdk.services.s3.S3BaseClientBuilder;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static org.apache.hadoop.fs.s3a.Constants.AWS_S3_ACCESS_GRANTS_ENABLED;
+
+
+/**
+ * Test S3 Access Grants configurations.
+ */
+public class TestS3AccessGrantConfiguration extends AbstractHadoopTestBase {
+    /**
+     * This credential provider will be attached to any client
+     * that has been configured with the S3 Access Grants plugin.
+     * {@link software.amazon.awssdk.s3accessgrants.plugin.S3AccessGrantsPlugin}.
+     */
+    public static final String S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS =
+            S3AccessGrantsIdentityProvider.class.getName();
+
+    @Test
+    public void testS3AccessGrantsEnabled() throws IOException, URISyntaxException {
+        // Feature is explicitly enabled
+        AwsClient s3AsyncClient = getAwsClient(createConfig(true), true);
+        Assert.assertEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3AsyncClient));
+
+        AwsClient s3Client = getAwsClient(createConfig(true), false);
+        Assert.assertEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3Client));
+    }
+
+    @Test
+    public void testS3AccessGrantsDisabled() throws IOException, URISyntaxException {
+        // Disabled by default
+        AwsClient s3AsyncDefaultClient = getAwsClient(new Configuration(), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3AsyncDefaultClient));
+
+        AwsClient s3DefaultClient = getAwsClient(new Configuration(), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3DefaultClient));
+
+        // Disabled if explicitly set
+        AwsClient s3AsyncExplicitlyDisabledClient = getAwsClient(createConfig(false), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3AsyncExplicitlyDisabledClient));
+
+        AwsClient s3ExplicitlyDisabledClient = getAwsClient(createConfig(false), true);
+        Assert.assertNotEquals(
+                S3_ACCESS_GRANTS_EXPECTED_CREDENTIAL_PROVIDER_CLASS,
+                getCredentialProviderName(s3ExplicitlyDisabledClient));
+    }
+
+    private Configuration createConfig(boolean s3agEnabled) {
```

Comment:
I think you'll need to do `removeBaseAndBucketOverrides` here before setting the value.

and is there a way to check for if the IAM fallback is set on the client?

Reply:
> I think you'll need to do removeBaseAndBucketOverrides here before setting the value.

I'm not sure about this because I'm starting a new Hadoop Configuration object itself rather than the `createConfiguration` methods that we use from the S3ATestUtils. In the end, I don't think it matters - because as long as we set the S3 Access Grants properties, that's all that matters to us for the purpose of this test, no?

> and is there a way to check for if the IAM fallback is set on the client?
Unfortunately not :( Did a lot of digging but in short, the plugins are "applied" to the client. When we apply the S3 Access Grants plugin on the S3 clients, we get the following identity provider set as the Credential Provider for this client: `S3AccessGrantsIdentityProvider`. And in the case of the fallback, the fallback flag is only set on the `S3AccessGrantsIdentityProvider` class but as a private variable that we cannot access.